File: src/main/java/com/hjj/apiserver/controller/ExceptionControllerAdvice.java
package com.hjj.apiserver.controller;


import com.hjj.apiserver.common.ApiError_Java;
import com.hjj.apiserver.common.ApiResponse_Java;
import com.hjj.apiserver.common.ApiUtils;
import com.hjj.apiserver.common.exception.AlreadyExistedUserException;
import com.hjj.apiserver.common.exception.ExistedSocialUserException;
import com.hjj.apiserver.common.exception.UserNotFoundException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import javax.servlet.http.HttpServletRequest;
import java.util.stream.Collectors;

@RestControllerAdvice
@Slf4j
public class ExceptionControllerAdvice {

    @ExceptionHandler(UserNotFoundException.class)
    @ResponseStatus(HttpStatus.OK)
    protected ApiResponse_Java userNotFoundException(HttpServletRequest request, UserNotFoundException e) {
        return ApiUtils.error(ApiError_Java.ErrCode.ERR_CODE0001.getMsg(), ApiError_Java.ErrCode.ERR_CODE0001);
    }

    @ExceptionHandler(AlreadyExistedUserException.class)
    @ResponseStatus(HttpStatus.OK)
    protected ApiResponse_Java alreadyExistedUserException(HttpServletRequest request, AlreadyExistedUserException e){
        return ApiUtils.error(ApiError_Java.ErrCode.ERR_CODE0006.getMsg(), ApiError_Java.ErrCode.ERR_CODE0006);
    }

    /* 소셜 로그인 계정이 있는데 일반계정 로그인 시도시 */
    @ExceptionHandler(ExistedSocialUserException.class)
    @ResponseStatus(HttpStatus.OK)
    protected ApiResponse_Java existedSocialUserException(HttpServletRequest request, ExistedSocialUserException e){
        return ApiUtils.error(ApiError_Java.ErrCode.ERR_CODE0007.getMsg(), ApiError_Java.ErrCode.ERR_CODE0007);
    }

    /* 패스워드가 일치하지 않은 경우 */
    @ExceptionHandler(BadCredentialsException.class)
    @ResponseStatus(HttpStatus.OK)
    protected ApiResponse_Java badCredentialsException(HttpServletRequest request, BadCredentialsException e){
        return ApiUtils.error(e.getMessage(), ApiError_Java.ErrCode.ERR_CODE9999);
    }


    @ExceptionHandler(MethodArgumentNotValidException.class)
    @ResponseStatus(HttpStatus.OK)
    protected ApiResponse_Java methodArgumentNotValidException(HttpServletRequest request, MethodArgumentNotValidException e){
        log.warn("Valid anotation 에러 통과하지 못함 url {}", request.getRequestURI());

        return ApiUtils.error(makeValidFailMessage(e.getBindingResult()), ApiError_Java.ErrCode.ERR_CODE9999);

    }


    public String makeValidFailMessage(BindingResult bindingResult){

        return bindingResult.getFieldErrors().stream().map(fieldError -> fieldError.getDefaultMessage())
                .collect(Collectors.joining(","));

    }
}

File: src/main/kotlin/com/hjj/apiserver/controller/ExceptionControllerAdvice.kt
package com.hjj.apiserver.controller

import com.hjj.apiserver.common.ApiError
import com.hjj.apiserver.common.ErrCode
import com.hjj.apiserver.common.exception.AccountBookNotFoundException
import com.hjj.apiserver.common.exception.AlreadyExistedUserException
import com.hjj.apiserver.common.exception.ExistedSocialUserException
import com.hjj.apiserver.common.exception.UserNotFoundException
import com.hjj.apiserver.util.logger
import org.slf4j.Logger
import org.springframework.http.HttpStatus
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.validation.BindingResult
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestControllerAdvice
import java.util.stream.Collectors
import jakarta.servlet.http.HttpServletRequest

@RestControllerAdvice
class ExceptionControllerAdvice{
    private val log: Logger = logger()


    @ExceptionHandler(UserNotFoundException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun userNotFoundException(request: HttpServletRequest): ApiError{
        return ApiError(ErrCode.ERR_CODE0001)
    }

    /* 소셜 로그인 계정이 있는데 일반계정 로그인 시도시 */
    @ExceptionHandler(ExistedSocialUserException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun existedSocialUserException(request: HttpServletRequest): ApiError{
        return ApiError(ErrCode.ERR_CODE0007)
    }

    /* 패스워드가 일치하지 않은 경우 */
    @ExceptionHandler(BadCredentialsException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun badCredentialsException(request: HttpServletRequest): ApiError{
        return ApiError(ErrCode.ERR_CODE0008)
    }

    @ExceptionHandler(AlreadyExistedUserException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun alreadyExistedUserException(request: HttpServletRequest): ApiError{
        return ApiError(ErrCode.ERR_CODE0006)
    }

    @ExceptionHandler(AccountBookNotFoundException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun accountBookNotFoundException(request: HttpServletRequest): ApiError{
        return ApiError(ErrCode.ERR_CODE0010)
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    @ResponseStatus(HttpStatus.BAD_REQUEST)
    protected fun methodArgumentNotValidException(request: HttpServletRequest, e: Exception): ApiError{
        return ApiError(ErrCode.ERR_CODE9999)
    }

    @ExceptionHandler(Exception::class)
    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
    protected fun basicException(request: HttpServletRequest, e: Exception): ApiError {
        log.error("[{}] Error Request: {}, ErrorInfo: {}", e.javaClass.name, request, e.printStackTrace())
        return ApiError(ErrCode.ERR_CODE9999)
    }

    fun makeValidFailMessage(bindingResult: BindingResult): String {
        return bindingResult.fieldErrors.stream().map { fieldError: FieldError -> fieldError.defaultMessage }
            .collect(Collectors.joining(","))
    }

}